### PR TITLE
[linstor] automatically fix symlinks for devices

### PR DESCRIPTION
**Pull request by kvaps**
## Description

Automatically fix symlinks for devices

Upstream PR https://github.com/LINBIT/linstor-server/pull/370

## Why do we need it, and what problem does it solve?

This change is workaround for specific set of issues, often related to udev,
which lead to the disappearance of symlinks for LVM devices on a working system.
These issues commonly occur during device resizing and deactivation,
causing LINSTOR exceptions during accessing DRBD super-block of volume.

## Why do we need it in the patch release (if we do)?
There is a customer who faces the problem quite often.

## What is the expected result?

When symlink to device is not exists, linstor will try to make it first.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: linstor
type: fix
summary: automatically fix symlinks for devices
impact_level: default
```

### CODE REVIEW

File: modules/041-linstor/images/linstor-server/patches/fix-symlinks.patch [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
++import com.linbit.linstor.layer.dmsetup.DmSetupUtils;
+ import com.linbit.linstor.layer.drbd.drbdstate.DrbdConnection;
+ import com.linbit.linstor.layer.drbd.drbdstate.DrbdEventPublisher;
+ import com.linbit.linstor.layer.drbd.drbdstate.DrbdResource;
+@@ -1143,6 +1144,18 @@ public class DrbdLayer implements DeviceLayer
+             metaDiskPath = drbdVlmData.getDataDevice();
+         }
+ 
++        if (!Files.exists(Paths.get(metaDiskPath))) {
++            errorReporter.logWarning("Device path %s does not exists, fixing symlink", metaDiskPath);
```
Comment (AleksZimin) [inline modules/041-linstor/images/linstor-server/patches/fix-symlinks.patch:143]:
```suggestion
+            errorReporter.logWarning("Device path %s does not exist, fixing symlink", metaDiskPath);
```

File: modules/041-linstor/images/linstor-server/patches/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+### Automatically fix symlinks for devices
+
+Introduce an additional handler that checks for the device path before each such modification.
+If the device is not found, it attempts to fix the symlink using dmsetup output.
+
+This change is workaround for specific set of issues, often related to udev,
+which lead to the disappearance of symlinks for LVM devices on a working system.
+These issues commonly manifest during device resizing and deactivation,
+causing LINSTOR expceptions when accessing DRBD super-block of volume.
```
Comment (AleksZimin) [inline modules/041-linstor/images/linstor-server/patches/README.md:11]:
```suggestion
causing LINSTOR exceptions when accessing DRBD super-block of volume.
```

File: modules/041-linstor/images/linstor-server/patches/fix-symlinks.patch [hunk 3]
```diff
@@ -0,0 +1,159 @@
+From 2f4e83bf4de23408a37621b10e6caf5f771c863c Mon Sep 17 00:00:00 2001
+From: Andrei Kvapil <kvapss@gmail.com>
+Date: Tue, 3 Oct 2023 13:32:21 +0200
+Subject: [PATCH] Automatically fix symlinks for devices
+
+Introduce an additional handler that checks for the device path before each such modification.
+If the device is not found, it attempts to fix the symlink using dmsetup output.
+
```
Comment (AleksZimin) [inline modules/041-linstor/images/linstor-server/patches/fix-symlinks.patch:8]:
```suggestion
```

File: modules/041-linstor/images/linstor-server/patches/fix-symlinks.patch [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+This change is workaround for specific set of issues, often related to udev,
+which lead to the disappearance of symlinks for LVM devices on a working system.
+These issues commonly occur during device resizing and deactivation,
+causing LINSTOR expceptions diring accessing DRBD super-block of volume.
```
Comment (AleksZimin) [inline modules/041-linstor/images/linstor-server/patches/fix-symlinks.patch:12]:
```suggestion
causing LINSTOR exceptions diring accessing DRBD super-block of volume.
```

**Review by AleksZimin — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [6143](https://github.com/deckhouse/deckhouse/pull/6143) to the branch [release-1.52](https://github.com/deckhouse/deckhouse/tree/release-1.52) successful!

**Pull request merged**